pyiceberg/table/__init__.py

```diff
@@ -1935,3 +2043,184 @@ def _generate_snapshot_id() -> int:
+
+    def _manifests(self) -> List[ManifestFile]:
+        manifests = []
+        deleted_entries = self._deleted_entries()
+
+        if self._added_data_files:
+            output_file_location = _new_manifest_path(location=self._table.location(), num=0, commit_uuid=self._commit_uuid)
+            with write_manifest(
+                format_version=self._table.format_version,
+                spec=self._table.spec(),
+                schema=self._table.schema(),
+                output_file=self._table.io.new_output(output_file_location),
+                snapshot_id=self._snapshot_id,
+            ) as writer:
+                for data_file in self._added_data_files:
+                    writer.add_entry(
+                        ManifestEntry(
+                            status=ManifestEntryStatus.ADDED,
+                            snapshot_id=self._snapshot_id,
+                            data_sequence_number=None,
+                            file_sequence_number=None,
+                            data_file=data_file,
+                        )
+                    )
+
+                for delete_entry in deleted_entries:
+                    writer.add_entry(delete_entry)
+
+            manifests.append(writer.to_manifest_file())
+
+        return manifests
+
+    def _summary(self) -> Dict[str, str]:
+        ssc = SnapshotSummaryCollector()
+
+        for data_file in self._added_data_files:
+            ssc.add_file(data_file=data_file)
+
+        return ssc.build()
+
+    def commit(self) -> Snapshot:
+        new_manifests = self._manifests()
+        next_sequence_number = self._table.next_sequence_number()
+
+        previous_snapshot = self._table.snapshot_by_id(self._parent_snapshot_id) if self._parent_snapshot_id is not None else None
+        summary = update_snapshot_summaries(
+            summary=Summary(operation=self._operation, **self._summary()),
+            previous_summary=previous_snapshot.summary if previous_snapshot is not None else None,
+            truncate_full_table=self._operation == Operation.OVERWRITE,
+        )
+
+        manifest_list_file_path = _generate_manifest_list_path(
+            location=self._table.location(), snapshot_id=self._snapshot_id, attempt=0, commit_uuid=self._commit_uuid
+        )
+        with write_manifest_list(
+            format_version=self._table.metadata.format_version,
+            output_file=self._table.io.new_output(manifest_list_file_path),
+            snapshot_id=self._snapshot_id,
+            parent_snapshot_id=self._parent_snapshot_id,
+            sequence_number=next_sequence_number,
+        ) as writer:
+            if self._operation == Operation.APPEND and previous_snapshot is not None:
+                # In case we want to append, just add the existing manifests
+                writer.add_manifests(previous_snapshot.manifests(io=self._table.io))
+            writer.add_manifests(new_manifests)
```

Comment:
Similar to the note above, I think it would be cleaner to have `_manifests` produce the complete set of manifests, not just the replacement ones. That method already relies on `_deleted_entries` to produce deletes, so it may as well also be responsible for checking whether to include the existing manifests.

Another option is to make `_manifests` produce just manifests for the appended files and handle deletes separately, but it looks like your approach here is to create just one manifest with both deletes and appends.

Reply:
Great suggestion. I've moved all the logic to `_manifests()`